sw/host/opentitanlib/src/transport/ti50emulator/emu.rs

```diff
@@ -214,58 +233,54 @@ impl EmulatorProcess {
-                        log::debug!("Wait for sub-process...");
-                        std::thread::sleep(TIMEOUT);
-                        retry += 1;
-                        if retry >= MAX_RETRY {
-                            return Err(EmuError::StartFailureCause(
-                                "Spawning Ti50Emulator sub-process timeout".to_string(),
-                            ));
-                        }
-                    }
-                    Err(_err) => {
-                        return Err(EmuError::RuntimeError(
-                            "Control socket io error".to_string(),
-                        ));
-                    }
-                }
-            }
-            Err(EmuError::StartFailureCause(
-                "Waiting for sub-process failed".to_string(),
-            ))
-        });
-
         log::info!("Spawning Ti50Emulator sub-process");
         log::info!("Command: {} {:?}", exec.display(), args_list);
-        let handle = Command::new(&exec)
+        let mut handle = Command::new(&exec)
             .args(args_list)
+            .stdin(Stdio::null())
+            .stdout(Stdio::piped())
             .spawn()
             .context("Could not spawn sub-process")?;
-        self.proc = Some(handle);
-        match ready_handle
-            .join()
-            .expect("Can't join control socket thread")
-        {
-            Ok(_) => Ok(()),
-            Err(err) => Err(err.into()),
+
+        log::info!("Waiting for sub-process start");
+        let child_stdout = handle.stdout.as_mut().expect("Failed to open stdout");
+        let mut buffer = Vec::new();
+        let mut timeout = TimeVal::microseconds(SPAWN_TIMEOUT.as_micros().try_into()?);
+        let mut fdset = select::FdSet::new();
+        fdset.insert(child_stdout.as_raw_fd());
+        loop {
+            // Select will subtract from the timeout, so that total time waited across all
+            // iterations of the loop is bounded.
+            if select::select(None, Some(&mut fdset), None, None, Some(&mut timeout))? == 0 {
+                // Timeout
+                return Err(EmuError::StartFailureCause(
+                    "Spawning Ti50Emulator sub-process timeout".to_string(),
+                )
+                .into());
+            }
+            let prev_len = buffer.len();
+            buffer.resize(prev_len + 256, 0u8);
+            let read_count = child_stdout.read(&mut buffer[prev_len..])?;
+            if read_count == 0 {
+                // Sub-process closed its stdout.  This can reasonably be assumed to mean
+                // that it has terminated.
+                let exit_status = handle.wait()?;
+                return Err(EmuError::StartFailureCause(format!(
+                    "Ti50Emulator sub-process exited with error: {}",
+                    exit_status
+                ))
+                .into());
+            }
+            buffer.resize(prev_len + read_count, 0u8);
```

Comment:
Is this just shortening the buffer in light of the read above?  If so, this can be removed because of `read_to_end`.

Reply:
Yes, this is shortening the buffer so that the length matches the number of bytes accumulated from `read()`s so far (except previous complete lines, which have been "drained" below).